Fix: validate schedule_expression for EventBridge Scheduler

anisaoshafi opened the pull request:
<!-- Please refer to the contribution guidelines before raising a PR: https://github.com/localstack/localstack/blob/master/docs/CONTRIBUTING.md -->

<!-- Why am I raising this PR? Add context such as related issues, PRs, or documentation. -->
## Motivation
Customer reported an issue for scheduler validation of schedule_expression  https://github.com/localstack/localstack/issues/9284.
The scheduler relies fully on moto for now, and validation is not implemented for `schedule_expression`.
Schedule_expression should support `at`, `cron` and `rate`.

<!-- What changes does this PR make? How does LocalStack behave differently now? -->
## Changes
Patched the `create_schedule` method in order to validate `schedule_expression` .
Reused some REGEX rules from events service.

<!-- Optional section: How to test these changes? -->
<!--
## Testing

-->

<!-- Optional section: What's left to do before it can be merged? -->
<!--
## TODO

What's left to do:

- [ ] ...
- [ ] ...
-->

bentsku commented on tests/aws/services/scheduler/test_scheduler.py:
```diff
@@ -58,3 +59,47 @@ def test_untag_resource(aws_client, events_scheduler_create_schedule_group, snap
+        "at(2021-12-31)",
+    ],
+)
+def tests_create_schedule_with_invalid_schedule_expression(
```
nice test! 👌

bentsku commented on localstack-core/localstack/services/scheduler/provider.py:
```diff
@@ -1,10 +1,39 @@
     pass
+
+
+def _validate_schedule_expression(schedule_expression) -> None:
```
nit: we could add a type hint for the `schedule_expression` argument

bentsku reviewed (commented):
LGTM! Nice port of your previous fix 👌 nice usage of the `patch` function. 

I only have a comment, it seems there code that cannot be executed, and that path is also untested. Once this is addressed, I believe the PR is ready to be merged! 🚀

anisaoshafi commented on localstack-core/localstack/services/scheduler/provider.py:
```diff
@@ -1,10 +1,39 @@
+
+
+def _validate_schedule_expression(schedule_expression) -> None:
+    if not schedule_expression:
+        raise ValidationException(
+            "Parameter(s) EventPattern or ScheduleExpression must be specified."
+        )
```
@bentsku sorry, I'm not following. Could you please rephrase this?

bentsku replied:
Sure, sorry! 
Before calling the `_validate_schedule_expression`, you do this:
`if schedule_expression := kwargs.get("schedule_expression"):`. 

Meaning that the `schedule_expression` parameter for `_validate_schedule_expression` cannot be false-ish, thus `if not schedule_expression:` can never be true. 

Also, this specific exception message `"Parameter(s) EventPattern or ScheduleExpression must be specified."` is not in the snapshot, so I figured it was not tested.

Maybe we could add a small test case for this? Or remove this validation?

anisaoshafi replied:
thanks, you're right. I'm going to double check it 🙏🏼

bentsku commented on tests/aws/services/scheduler/test_scheduler.py:
```diff
@@ -103,3 +103,46 @@ def tests_create_schedule_with_invalid_schedule_expression(
+):
+    rule_name = f"rule-{short_uid()}"
+
+    with pytest.raises(ParamValidationError) as e:
+        aws_client.scheduler.create_schedule(
+            Name=rule_name,
+            ScheduleExpression="",
+            FlexibleTimeWindow={
+                "MaximumWindowInMinutes": 4,
+                "Mode": "FLEXIBLE",
+            },
+            Target={
+                "Arn": f"arn:aws:lambda:{region_name}:{account_id}:function:dummy",
+                "RoleArn": f"arn:aws:iam::{account_id}:role/role-name",
+            },
+        )
+    snapshot.match("empty-schedule-expression", e.value)
```
`ParamValidationError` indicates that the error is coming from the client itself, and the request is not sent to AWS or LocalStack. We don't need to snapshot those kind of exception as they are dependent on the client only. 

To configure the client to not have validation, and send the invalid request against AWS, you can configure you client this way:
```python
from botocore.client import Config

def tests_create_schedule_with_empty_schedule_expression(
    aws_client_factory, region_name, account_id, snapshot
):
    scheduler_client = aws_client_factory(config=Config(parameter_validation=False)).scheduler
    rule_name = f"rule-{short_uid()}"

   with pytest.raises(ParamValidationError) as e:
        scheduler_client.create_schedule(
        [...]
```

This way, the client won't validate the request and will let you send the invalid request against AWS, and see the proper exception.

anisaoshafi replied:
understood, thanks for sharing this. I'll keep it in mind for the future and will proceed removing these two dummy tests.

anisaoshafi commented:
> LGTM! Nice port of your previous fix 👌 nice usage of the `patch` function.
> 
> I only have a comment, it seems there code that cannot be executed, and that path is also untested. Once this is addressed, I believe the PR is ready to be merged! 🚀


@bentsku Thanks for your comments. Could you please double check this when you have a moment? Thanks 🙌🏼

bentsku reviewed (approved):
LGTM! Thanks for addressing the comments. 

I think we could have kept the tests when the `schedule_expression` was missing, but with the client parameter validation disabled to see what AWS returned, however it's not fully in the scope of the issue, so that's alright, it was just a nice to have as you had already added it 😄 and it's pretty niche as most of the SDK will validate the input. 

Anyway, very nice fix and parametrized tests! 🚀💯

anisaoshafi commented:
> LGTM! Thanks for addressing the comments.
> 
> I think we could have kept the tests when the `schedule_expression` was missing, but with the client parameter validation disabled to see what AWS returned, however it's not fully in the scope of the issue, so that's alright, it was just a nice to have as you had already added it 😄 and it's pretty niche as most of the SDK will validate the input.
> 
> Anyway, very nice fix and parametrized tests! 🚀💯

Sorry I deleted everything. I'll merge this as is and if I come back to eventbridge might add that test 🙏🏼

bentsku commented:
> Sorry I deleted everything. I'll merge this as is and if I come back to eventbridge might add that test 🙏🏼

No worries! You already went through the effort, that was all 😄 but this PR is good as it is, so let's get this merged! 🤝

ridgey-dev commented:
I now get the error:

Invalid Schedule Expression at(2025-01-31T15:59:42)

with value:

`at(2025-01-31T15:59:42)`

Looking at AWS documentation [https://docs.aws.amazon.com/scheduler/latest/UserGuide/schedule-types.html#one-time](url) it's valid, but looking at the regex in the PR it's not?

anisaoshafi commented:
Hi @ridgey-dev. Thanks for reporting this 🙏🏼 
I double checked with aws and the date you passed seems valid.
> it's expecting d-m-Y instead of Y-m-d?

No, it's "Y-m-d", I'll fix the validation rule since I closed it too much.

ridgey-dev commented:
Thanks for your quick answer. Looking forward to your fix! Our pipeline that runs the tests on LocalStack is now failing because of this.

anisaoshafi commented:
Hi @ridgey-dev, thank you for your patience. We have implemented changes that should resolve your reported issue. Kindly update to the latest LocalStack container image and let us know if your issue has been resolved.

ridgey-dev commented:
It's working again. Thanks for your quick fix!
